[Internal] Refactor provider and related packages (#3940)
databricks/terraform-provider-databricks: CONTRIBUTING.md

After installing the necessary software for building provider from sources, you should be able to run `make coverage` to run the tests and see the coverage.

## Package organization for Providers
We are migrating the resource from SDKv2 to Plugin Framework provider and hence both of them exist in the codebase. For uniform code convention, readability and development, they are defined under internal directory under root as: 

> - `common/provider`: This package contains the changes common to both the providers example provider factory and tests

Reviewer: There is no `internal/common/provider` package in this PR or on main. Are you planning to introduce this, or is this a stale comment?

Author: There is internal/common/provider in the PR:
<img width="297" alt="image" src="https://github.com/user-attachments/assets/bcb2cd7e-a580-44ef-9900-c8d30acdb66b">

Maybe I am getting this wrong, do you mean imports for common provider?
We use that in main.go

`"github.com/databricks/terraform-provider-databricks/internal/common/provider"`

  - imports are declared as: 

>   	- `pluginframeworkprovider` "github.com/databricks/terraform-provider-databricks/internal/pluginframework/provider"
> 	  - `sdkv2provider` "github.com/databricks/terraform-provider-databricks/internal/sdkv2/provider"

Reviewer: Small suggestion just to shorten the length of the renamed import slightly
```suggestion
  	- `fwprovider` "github.com/databricks/terraform-provider-databricks/internal/pluginframework/provider"
	  - `sdkv2provider` "github.com/databricks/terraform-provider-databricks/internal/sdkv2/provider"
```

- `sdkv2/provider`: This package contains the changes related to only sdkv2 provider and shouldn't depend on pluginframework provider

> - `pluginframework/provider`: This package contains the changes related to only plugin framework provider and shouldn't depend on sdkv2

Reviewer 2: I think it would be good to start building the habit of documenting our packages, even internal ones. Having these instructions next to the code is located will increase the likelihood of people reading them.

See "Packages" section: https://tip.golang.org/doc/comment#packages

> - `pluginframework/provider`: This package contains the changes related to only plugin framework provider and shouldn't depend on sdkv2

Reviewer 2: Have we consider a different structure where all these will be subpackages of a `providers` package?

```
- providers/
  - common/
  - sdkv2/
  - pluginfw/
``` 

This is a little more idiomatic, makes it clear that all of this is provider related, and solve the need for renaming packages.

Author: Agree, that's better for organisation. I will update the structure.

> Note: To avoid circular dependency and make code easy to read, any common changes for both provider will go into interna/common/provider package.

Reviewer 2: Phrasing nit + fix a couple of typos.

```suggestion
Note: Common dependencies for both providers will go into the internal/common/provider package.
```
